src/types/mod.rs

```diff
@@ -86,6 +86,21 @@ pub mod iter {
     pub use super::tuple::{BorrowedTupleIterator, BoundTupleIterator, PyTupleIterator};
 }
 
+/// Python objects that have a base type.
+///
+/// This marks types that can be upcast into a [`PyAny`] and used in its place.
+/// This essentially includes every Python object except [`PyAny`] itself.
+///
+/// This is used to provide the [`AsRef<Bound<'_, PyAny>`](std::convert::AsRef)
+/// and [`Deref<Target = Bound<'_, PyAny>`](std::ops::Deref) implementations for
+/// [`Bound<'_, T>`](crate::Bound)
+///
+/// Users should not need to implement this trait directly. It's implementation
+/// is provided by the [`#[pyclass]`](macro@crate::pyclass) attribute.
+pub trait HasPyBaseType {
```

Comment:
Could this functionality also provided by `PyTypeCheck` or `PyTypeInfo`?

Reply:
(It is nice to have targetted minimalistic traits, but we probably also would to avoid making our infrastructure traits unnecessarily complicated and hard to keep in one's head.)

Reply:
The problem is that these are also implemented for `PyAny`, so we run into the same problem as without any bound